File: src/main/kotlin/space/davids_digital/kiri/rest/dto/ToolDto.kt
package space.davids_digital.kiri.rest.dto

data class ToolDto(
    val fullName: String,
    val description: String?,
    val parameters: ToolParameterValueDto
)

sealed interface ToolParameterValueDto {
    val type: String

    data class ObjectValue(
        val description: String?,
        val properties: Map<String, ToolParameterValueDto>,
        val required: List<String>
    ) : ToolParameterValueDto {
        override val type = "object"
    }

    data class ArrayValue(
        val description: String?,
        val items: ToolParameterValueDto
    ) : ToolParameterValueDto {
        override val type = "array"
    }

    data class StringValue(
        val description: String?,
        val enum: List<String>?
    ) : ToolParameterValueDto {
        override val type = "string"
    }

    data class NumberValue(
        val description: String?
    ) : ToolParameterValueDto {
        override val type = "number"
    }

    data class BooleanValue(
        val description: String?
    ) : ToolParameterValueDto {
        override val type = "boolean"
    }
}

data class ToolExecuteRequest(
    val toolName: String,
    val input: ToolInputDto
)

File: admin-ui/lib/api/types/ToolDto.ts
import {ToolInputDto} from "./FrameDto";

export type ToolDto = {
    fullName: string;
    description: string | null;
    parameters: ToolParameterValueDto;
};

export type ToolParameterValueDto =
    | ToolParameterObjectValue
    | ToolParameterArrayValue
    | ToolParameterStringValue
    | ToolParameterNumberValue
    | ToolParameterBooleanValue;

export type ToolParameterObjectValue = {
    type: "object";
    description: string | null;
    properties: { [key: string]: ToolParameterValueDto };
    required: string[];
};

export type ToolParameterArrayValue = {
    type: "array";
    description: string | null;
    items: ToolParameterValueDto;
};

export type ToolParameterStringValue = {
    type: "string";
    description: string | null;
    enum: string[] | null;
};

export type ToolParameterNumberValue = {
    type: "number";
    description: string | null;
};

export type ToolParameterBooleanValue = {
    type: "boolean";
    description: string | null;
};

export type ToolExecuteRequest = {
    toolName: string;
    input: ToolInputDto;
};
